### PR TITLE
Extend ValueWriter/ValueReader with array based operations

### PR DESCRIPTION
**Pull request by jwillemsen**
Extend ValueWriter/Reader with array based operation, currently these are not usable with large arrays/sequences because there is a callback to user code for each element of the array
* Extend ValueWriter/Reader with array based operations, default implementation iterates through the array and invokes the operations as in the past
* Update the IDL compiler to use the new array based operations, reduces the amount of code generated
* Extended and add new unit tests to test more of the generated code, previously there was no test for the vread/vwrite at all

### CODE REVIEW

**Comment by jwillemsen** [comment]
Build & Test / build_w19_re_o1p1_sec failed with the error below, looks unrelated to my changes

LINK : fatal error LNK1181: cannot open input file 'ssleay32.lib' [D:\a\OpenDDS\OpenDDS\OpenDDS\dds\DCPS\security\OpenDDS_Security.vcxproj]

File: dds/DCPS/ValueReader.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+bool
+ValueReader::read_float128(long double& value)
+{
+  ACE_CDR::LongDouble ld;
+  if (!read_float128(ld)) {
+    return false;
+  }
+  value = ld;
+  return true; }
```
Comment (jrw972) [inline dds/DCPS/ValueReader.cpp:23]:
```suggestion
  return true;
}
```

File: dds/DCPS/ValueReader.cpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+#include "DCPS/DdsDcps_pch.h" //Only the _pch include should start with DCPS/
+
+#include "ValueReader.h"
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+
+namespace OpenDDS {
+namespace DCPS {
+
+bool
```
Comment (jrw972) [inline dds/DCPS/ValueReader.cpp:15]: Conditional on `#ifdef NONNATIVE_LONGDOUBLE`

**Review by jrw972 — CHANGES_REQUESTED** [review]

File: dds/DCPS/ValueReader.cpp (hunk 2 above)
Comment (jwillemsen) [reply dds/DCPS/ValueReader.cpp:16]: Done

**Review by jrw972 — APPROVED** [review]

**Comment by mitza-oci** [comment]
> 
> 
> Build & Test / build_w19_re_o1p1_sec failed with the error below, looks unrelated to my changes
> 
> ```
> LINK : fatal error LNK1181: cannot open input file 'ssleay32.lib' [D:\a\OpenDDS\OpenDDS\OpenDDS\dds\DCPS\security\OpenDDS_Security.vcxproj]
> ```

Yes, there's a GitHub Actions configuration issue impacting all builds.  We'll look into it.

**Comment by mitza-oci** [comment]
> 
> 
> > Build & Test / build_w19_re_o1p1_sec failed with the error below, looks unrelated to my changes
> > ```
> > LINK : fatal error LNK1181: cannot open input file 'ssleay32.lib' [D:\a\OpenDDS\OpenDDS\OpenDDS\dds\DCPS\security\OpenDDS_Security.vcxproj]
> > ```
> 
> Yes, there's a GitHub Actions configuration issue impacting all builds. We'll look into it.

Attempting to diagnose & fix here: https://github.com/objectcomputing/OpenDDS/pull/3143

File: tests/unit-tests/dds/DCPS/JsonValueWriter.cpp [hunk 3]
```diff
@@ -12,6 +12,18 @@
 using namespace rapidjson;
 using namespace OpenDDS::DCPS;
 
+class TestWriter : public JsonValueWriter<>
+{
+public:
+  TestWriter () : elements_ (0) {};
```
Comment (mitza-oci) [inline tests/unit-tests/dds/DCPS/JsonValueWriter.cpp:18]:
```suggestion
  TestWriter () : elements_ (0) {}
```
There's a few of these.

File: tests/unit-tests/dds/DCPS/JsonValueWriter.cpp [hunk 4]
```diff
@@ -159,9 +171,34 @@ TEST(dds_DCPS_JsonValueWriter, end_element)
...
+  jvw.begin_element(1);
+  jvw.write_int16(6);
+  jvw.end_element();
+  jvw.end_sequence();
+  EXPECT_STREQ(jvw.buffer().GetString(), "[5,6]");
+}
+
+TEST(dds_DCPS_JsonValueWriter, complete_sequence_write_array)
+{
+  ACE_CDR::Short const i[2] = {5, 6};
```
Comment (mitza-oci) [inline tests/unit-tests/dds/DCPS/JsonValueWriter.cpp:190]:
```suggestion
  const ACE_CDR::Short i[]= {5, 6};
```

File: dds/DCPS/ValueReader.h [hunk 5]
```diff
@@ -100,7 +100,9 @@ class ListEnumHelper : public EnumHelper {
   const Pair* pairs_;
 };
 
-struct ValueReader {
+class OpenDDS_Dcps_Export ValueReader {
+public:
```
Comment (mitza-oci) [inline dds/DCPS/ValueReader.h:104]: Does it have a class invariant?  Is anything private?  If neither is true, isn't `struct` clear and sufficient?

File: tests/DCPS/Observer/TestObserver.h [hunk 6]
```diff
@@ -60,7 +60,8 @@ class TestObserver : public virtual OpenDDS::DCPS::Observer
 
   template<typename Qos>
   static std::string qos_str(const Qos& qos) {
-    std::ostringstream o; o << " qos.user_data: ";
+    std::ostringstream o;
+    o << " qos.user_data: ";
```
Comment (mitza-oci) [inline tests/DCPS/Observer/TestObserver.h:64]: It would be preferrable to have unrelated changes in a separate PR from one that adds a new feature.

File: dds/DCPS/ValueReader.h (hunk 5 above)
Comment (jwillemsen) [reply dds/DCPS/ValueReader.h:104]: I personally prefer `class`

Comment (mitza-oci) [reply dds/DCPS/ValueReader.h:104]: Not at all a requirement... just to give more background here, we should try to follow C++ Core Guidelines when there isn't a project-specific guideline.  In this case there may not be a clear consensus.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c122-use-abstract-classes-as-interfaces-when-complete-separation-of-interface-and-implementation-is-needed

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c2-use-class-if-the-class-has-an-invariant-use-struct-if-the-data-members-can-vary-independently

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c8-use-class-rather-than-struct-if-any-member-is-non-public
